### PR TITLE
#1340 Allow Import of Dependent Permissions

### PR DESCRIPTION
**Pull request by tadean**
This PR alters the 9.3 permission changes to allow for two kinds of permissions to be assigned on the fly; provided permissions and dependent permissions. Previously there was no issue with provided permissions, but in the case of dependent permissions they were not being assigned (eg. `edit any book content` being provided by `node`, but actually existing because of `book`). In this PR, we attempt to additionally detect if a module contains a piece of configuration the permission depends on, which we take to assume that it should have a chance for quickstart to set its permissiions.

Additionally, we alllow the profile a chance to install permissions for its dependent modules.

## Related issues
#1340 
#1181 

## How to test
Test that bundle-based permissions are now enabled correctly.

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php [hunk 1]
```diff
@@ -167,12 +195,9 @@ protected function trimPermissions(array $config) {
    *   An array of permission definitions.
    */
   protected function getPermissionDefinitions() {
-    if (empty($this->permissionDefinitions)) {
-      // @todo Use injection on user.permissions.
-      // @phpstan-ignore-next-line
-      $this->permissionDefinitions = \Drupal::service('user.permissions')->getPermissions();
-    }
-    return $this->permissionDefinitions;
+    // @todo Use injection on user.permissions. and add caching.
```
Comment (tadean) [inline modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php:198]: Something seems to be going on with the caching that was here previously. With it enabled, it was receiving the wrong results for some kinds of permissions...investigation needed.

**Comment by trackleft** [comment]
Tested with quick node clone and it worked
image

File: modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php (hunk 1 above)
Comment (joeparsons) [reply modules/custom/az_core/src/Plugin/ConfigProvider/QuickstartConfigProvider.php:198]: > Something seems to be going on with the caching that was here previously. With it enabled, it was receiving the wrong results for some kinds of permissions...investigation needed.

The core patch that adds static caching to the PermissionsHandler service does this in `hook_modules_installed()`:
```
// Reset the user.permissions service to reset statically cached
// permissions.
\Drupal::getContainer()->set('user.permissions', NULL);
```

We may need to do something similar to reset the statically cached permissions in this plugin class, although I'm not sure what the best way to do so would be.

**Comment by trackleft** [comment]
> Are they supposed to have the ability to assign actual admins?

They shouldn't be

**Comment by joeparsons** [comment]
> > Looks like entity bundle permissions that were missing before exist now.
> > Others are still missing (probably expected):
> > 
> > * Role delegation permissions
> > * Full HTML text format permissions
> 
> Aren't these the role delegation permissions that were missing? role

Yes.  Those are the ones.  I'll try rebuilding my local env and try again.

**Comment by joeparsons** [comment]
> Are they supposed to have the ability to assign actual admins?

No.  I was seeing none of them enabled in my testing but I was also experimenting with re-exporting config so I might have messed something up.

**Comment by tadean** [comment]
Not sure if this would be related, but those permissions will only show up on a new drupal install, if they'd missed their opportunity to be set the first time (since their module gets installed before the profile).

**Review by joeparsons — APPROVED** [review]
Please disregard my earlier comments/questions about missing permissions.  When I initially cloned the Quickstart repo to test this locally, I foolishly did `git checkout -b issue/1340` instead of `git checkout issue/1340` 🙃.

After checking out the correct branch instead of creating a new one, I was able to verify that all expected permissions existed.  I also did several experiments with re-exporting the roles after installing optional core/contrib modules and found that all permissions in our roles are valid (tied to a module provided by Quickstart) except for 3 that I created a separate issue (#1581) and PR (#1582) for.

I think I'd be OK with merging this as-is so we can expedite creating a stable 2.3.x release based on Drupal 9.3.x ASAP but we'll probably want to create a follow-up issue to investigate whether anything can be done to reduce the performance impact of these changes:

<table>
<tr>
	<td>Task
	<td>main (#1582)
	<td>This PR (latest revision)
<tr>
	<td>Install Quickstart
	<td>2 minutes 11 seconds 72 milliseconds
	<td>2 minutes 56 seconds 153 milliseconds
<tr>
	<td>PHPunit tests
	<td>16 minutes 41 seconds 709 milliseconds
	<td>23 minutes 10 seconds 91 milliseconds
<tr>
	<td>Migration tests
	<td>4 minutes 28 seconds 43 milliseconds
	<td>5 minutes 18 seconds 618 milliseconds
</table>

**Comment by trackleft** [comment]
@tadean mentioned a follow-up task for tracking site-owner selected permissions in a config file we could use to determine the site owner's intended permissions so we do not ever override those.

**Comment by joeparsons** [comment]
Follow-up issues:

- #1595
- #1596
- #1597

**Pull request merged**